### DISCUSSION TITLE
Next.js 13 App Router + Chakra UI = FOUC (flash of unstyled content)

### DISCUSSION THREAD

**Discussion by Hulkstance** [chakra-ui/chakra-ui#8098] [General]
I've been going through the latest Next.js 13 (with /app folder) changes and created a simple application. When I refresh the page:
- The screen flashes white for a second before showing the dark theme.
- The internationalization and font settings are also back to default for a blink of an eye before applying the saved preferences.

How do you guys solve that _FOUC (flash of unstyled content)_? I'm aware that this happens because the Chakra UI stylesheets from `<ChakraProvider />{:tsx}` are not loaded server-side but only client-side.

## package.json

```json
{
  "name": "logistics",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start",
    "lint": "next lint",
    "format": "prettier --write ."
  },
  "dependencies": {
    "@chakra-ui/icons": "^2.1.1",
    "@chakra-ui/next-js": "^2.1.5",
    "@chakra-ui/react": "^2.8.1",
    "@emotion/react": "^11.11.1",
    "@emotion/styled": "^11.11.0",
    "framer-motion": "^10.16.4",
    "next": "14.0.0",
    "next-intl": "^2.21.0",
    "react": "^18",
    "react-dom": "^18"
  },
  "devDependencies": {
    "@types/node": "^20",
    "@types/react": "^18",
    "@types/react-dom": "^18",
    "eslint": "^8",
    "eslint-config-next": "14.0.0",
    "eslint-config-prettier": "^9.0.0",
    "eslint-plugin-prettier": "^5.0.1",
    "prettier": "^3.0.3",
    "typescript": "^5"
  }
}
```

**Comment by dev-aly3n** [comment]
same experience with next 14 here.
packages:
    "@chakra-ui/**next-js**": "**^2.2.0**",
    "@chakra-ui/**react**": "**2.7.1**",
    "@emotion/**react**": "^**11.11.1**",
    "@emotion/**styled**": "^**11.11.0**",
    "framer-motion": "^**10.16.16**",
    "**next**": "^**14.0.3**",
    "**react**": "^**18.2.0**",
    "react-dom": "^18.2.0"

**Comment by TerraNibble** [comment]
I have finally found a working solution to this issue on NextJS 14 by manually implementing a cookie based `colorModeManager` along with a `ColorModeScript`. The cookie then can be received server-side and used for SSR prior to hydration.

Here is my solution...

**layout.tsx**
```typescript
import { ThemeProvider } from "@/providers/ThemeProvider";
import { cookies } from "next/headers";

export default function RootLayout({
  children,
}: {
  children: React.ReactNode;
}) {
  const cookiesList = cookies();
  const colorMode = cookiesList.get("chakra-ui-color-mode");

  return (
    <html lang="en>
      <link rel="icon" href="/favicon.ico" sizes="any" />
      <body>
        <ThemeProvider colorMode={colorMode?.value}>
           <main className={styles.main}>{children}</main>
        </ThemeProvider>
      </body>
    </html>
  );
}
```

**ThemeProvider.tsx**
```typescript
"use client";

import { ChakraProvider, ColorModeScript } from "@chakra-ui/react";
import Theme from "@/lib/theme";
import React from "react";
import { CacheProvider } from "@chakra-ui/next-js";
import { setCookie } from "cookies-next";

export const ThemeProvider = ({
  colorMode,
  children,
}: {
  colorMode?: any;
  children: React.ReactNode;
}) => {
  return (
    <CacheProvider>
      <ChakraProvider
        colorModeManager={{
          type: "cookie",
          ssr: true,
          get: (init) => colorMode ?? init,
          set: (value) => {
            setCookie("chakra-ui-color-mode", value);
          },
        }}
        theme={Theme}
      >
        <ColorModeScript
          initialColorMode={Theme.initialColorMode}
          type="cookie"
        />
        {children}
      </ChakraProvider>
    </CacheProvider>
  );
};
```

**Reply by leandrolavore** [reply]
did you get the hydration warning error with this approach? I'm still getting a miss-match between the classes in the body tag on Server side and client side rendering (body tag does not get the `chakra-ui-dark` class on the server, just on the client)

**Reply by ghost** [reply]
I am facing the same issue here. Even with suppressHydrationWarnings prop set
